refactor with max frequency per row

JesperDramsch opened the pull request:
Refactor of column names to make code more legible and easily changeable. Corrected column number for deletion days.

Add Column for max update frequency per row as suggested in #21 

This PR only works with the following sheet:
![image](https://user-images.githubusercontent.com/2620316/74692538-4e8bb100-51f0-11ea-9f32-cb5e76b5e540.png)

JesperDramsch commented:
This way the update frequency for the sheet can be a high number like every hour, but each row is only updated according to column 3.

Works without values in column 3 and 4.

victorjzsun commented on sheetScript.gs:
```diff
@@ -66,91 +70,98 @@ function updatePlaylists(sheet) {
+	var freqDate = new Date(lastTimestamp);
+	var dateDiff = Date.now() - freqDate;
+	var nextTime = sheet.getRange(iRow + 1, reservedColumnFrequency).getValue()  * 36e5;
+	if (dateDiff <= nextTime && nextTime) {
```
Check nextTime first before comparison with dateDiff

JesperDramsch replied:
Why?

victorjzsun replied:
You're checking if nextTime exists (isn't an empty string) here right? Wouldn't it be better to check for existence before comparing it with dateDiff?

JesperDramsch replied:
I'll change it, but I don't think it matters, as an empty cell evaluates to 0.

victorjzsun commented on sheetScript.gs:
```diff
@@ -66,91 +70,98 @@ function updatePlaylists(sheet) {
+	    /// ...get channels...
+	    var channelIds = [];
+	    var playlistIds = [];
+	    Logger.log("Row: " + (iRow+1))
```
Move this line higher up so that user knows which row was skipped from log on line 87

JesperDramsch replied:
Good point.

victorjzsun commented on sheetScript.gs:
```diff
@@ -66,91 +70,98 @@ function updatePlaylists(sheet) {
-        [].push.apply(newVideoIds, videoIds);
-      }
+	
+	// Check if it's time to update already
```
Convert tabs to spaces for consistency

JesperDramsch replied:
I was wondering about this. Done

victorjzsun commented on sheetScript.gs:
```diff
@@ -25,12 +34,10 @@ function doGet(e) {
     function (min, row, index) {
-      if (row[reservedTimestampColumn].length != 0 && row[reservedTimestampColumn] < min[1]) {
-        return [index, row[reservedTimestampColumn]]
+      if (row[reservedColumnPlaylist].length != 0 && row[reservedColumnPlaylist] < min[1]) {
```
Wouldn't `reservedColumnTimestamp - 1` be more readable?

JesperDramsch replied:
That assumes you'll never change the order of the columns again. Considering that the code contained multiple mistakes with constant offsets (reservedTimestampColumn +1 +2 +3)  I recommend making these explicit at the beginning of the script.

victorjzsun replied:
Do whatever you think makes sense, and allows future developers to understand what's going on.

victorjzsun commented on sheetScript.gs:
```diff
@@ -66,91 +70,98 @@ function updatePlaylists(sheet) {
+	      }
+	      
+	      /// ...delete old vidoes in playlist
+	      var daysBack = data[iRow][reservedColumnDeleteDays];
```
One-off error since this is array index?

JesperDramsch replied:
In the Row or Column?

victorjzsun replied:
column

JesperDramsch replied:
There was an off by one error before, that I corrected. It was still pointing to the old second column.

JesperDramsch replied:
Or am I mixing things up here?

victorjzsun replied:
All of your constants are sheet column indexes, and any data array index should be `<constant> - 1` (like the `reservedColumnTimestamp - 1` = `reservedColumnPlaylist` indexing in findNextRow)

JesperDramsch replied:
I think you're right I apologize. Let me try and fix it.

JesperDramsch commented:
Okay, I changed the `getRange` codes to a call of the data array and gave an explanation in the comments with the variable to make it easier for new devs. Changed everything to the correct 0-index (sorry again). 

Now the only "off by X" correction has to be made when writing to the sheet, which is explained in the sheet.

Fixed a bug with a variable name `playlistId` vs `plalistID` that came up from time to time.

Addressed all commens and implemented those changes.

JesperDramsch commented:
Thanks for the thorough review!

victorjzsun commented on sheetScript.gs:
```diff
@@ -63,94 +70,101 @@ function updatePlaylists(sheet) {
+  var nextTime = data[iRow][reservedColumnFrequency]  * MILLIS_PER_HOUR;
+  if (nextTime && dateDiff <= nextTime) {
+    Logger.log("Skipped: Not time yet");
+  } else {
```
These lines should be 6-space indented for consistency with code above and below

victorjzsun commented on sheetScript.gs:
```diff
@@ -11,6 +11,17 @@ var debugFlag_dontUpdateTimestamp = false;
+
+// Define reserved Rows and Columns (zero-based)
+var reservedTableRows = 2;        // Start of the range of the PlaylistID+ChannelID data
+var reservedTableColumns = 4;     // Start of the range of the ChannelID data
```
These were zero-based before, so should stay as
```
var reservedTableRows = 3;        // Start of the range of the PlaylistID+ChannelID data
var reservedTableColumns = 5;     // Start of the range of the ChannelID data
```
We want reservedTableRows to be row 4 of the sheet, and reservedTable Columns to be column F of the sheet, the start of user data

JesperDramsch commented:
@victorjzsun All points addressed.
